app/com/linkedin/drelephant/tuning/engine/SparkHBTParamRecommender.java

```diff
@@ -0,0 +1,326 @@
+    return maxPeakJVMUsedMemory / lastRunExecutorCore;
+  }
+
+  private Long getMaxPeakUnifiedMemoryPerCore() {
+    return maxPeakUnifiedMemory / lastRunExecutorCore;
+  }
+
+  /**
+   * Suggest executor memory and core We calculate per core peak JVM used memory. Then we start from
+   * max executor core and compute memory requirement for max executor core If this is below
+   * threshold of max executor core then good, otherwise reduce the core and check memory
+   * requirement.
+   */
+  private void suggestExecutorMemoryCore() {
+    boolean validSuggestion = false;
+    long maxPeakJVMUsedMemoryPerCore = getMaxPeakJVMUsedMemoryPerCore();
+    for (int core = MAX_EXECUTOR_CORE; core > 0; core--) {
+      long currSuggestedMemory = suggestExecutorMemory(maxPeakJVMUsedMemoryPerCore, core);
+      if (currSuggestedMemory < MAX_EXECUTOR_MEMORY || core == 1) {
+        suggestedExecutorMemory = currSuggestedMemory;
+        suggestedCore = core;
+        validSuggestion = true;
+        break;
+      }
+    }
+    if (!validSuggestion) {
+      suggestedExecutorMemory = lastRunExecutorMemory;
+      suggestedCore = lastRunExecutorCore;
+    }
+    suggestedExecutorMemory = getRoundedExecutorMemory();
+  }
+
+  /**
+   * This method is to get the rounded executor memory. We don't want to round executor memory to
+   * multiple of 1GB as container size is decided by executor memory + memoryOverhead. So round the
+   * executor memory + memoryOverhead and then distribute extra memory to executor memory
+   *
+   * @return rounded executor memory
+   */
+  private long getRoundedExecutorMemory() {
+    long memoryOverhead = lastRunExecutorMemoryOverhead;
+    if (lastRunExecutorMemoryOverhead == 0) {
+      memoryOverhead = Math.max(suggestedExecutorMemory / 10, 384);
+    }
+
+    long roundedContainerSize = getRoundedContainerSize(suggestedExecutorMemory + memoryOverhead + FileUtils.ONE_MB);
+    long executorMemory =
+        (long) (suggestedExecutorMemory + (roundedContainerSize - suggestedExecutorMemory - memoryOverhead - FileUtils.ONE_MB) * 0.9);
+
+    return executorMemory;
+  }
+
+  /**
+   * This method is to get the rounded executor memory. We don't want to round executor memory to
+   * multiple of 1GB as container size is decided by executor memory + memoryOverhead. So round the
+   * executor memory + memoryOverhead and then distribute extra memory to executor memory
+   *
+   * @return rounded driver memory
+   */
+  private long getRoundedDriverMemory() {
+    long memoryOverhead = lastRunDriverMemoryOverhead;
+    if (lastRunDriverMemoryOverhead == 0) {
+      memoryOverhead = Math.max(suggestedDriverMemory / 10, 384);
+    }
+
+    long roundedContainerSize = getRoundedContainerSize(suggestedDriverMemory + memoryOverhead + FileUtils.ONE_MB);
+    long driverMemory =
+        (long) (suggestedDriverMemory + (roundedContainerSize - suggestedDriverMemory - memoryOverhead - FileUtils.ONE_MB) * 0.9);
+
+    return driverMemory;
+  }
```

Comment:
Can we combine these two methods , getRoundedExecutorMemory and getRoundedDriverMemory ?

Reply:
I think that's fine. Both are using different parameters. Currently logic is similar but going forward it may be different.